pandas/tests/tseries/offsets/test_offsets.py

```diff
@@ -792,6 +784,41 @@ def test_tick_normalize_raises(tick_classes):
         cls(n=3, normalize=True)
 
 
+@pytest.mark.parametrize(
+    "cases",
+    [
+        ("nanoseconds", 1, Timestamp("1970-01-01 00:00:00.000000001")),
+        ("nanoseconds", 5, Timestamp("1970-01-01 00:00:00.000000005")),
+        ("nanoseconds", -1, Timestamp("1969-12-31 23:59:59.999999999")),
+        ("microseconds", 1, Timestamp("1970-01-01 00:00:00.000001")),
+        ("microseconds", -1, Timestamp("1969-12-31 23:59:59.999999")),
+        ("seconds", 1, Timestamp("1970-01-01 00:00:01")),
+        ("seconds", -1, Timestamp("1969-12-31 23:59:59")),
+        ("minutes", 1, Timestamp("1970-01-01 00:01:00")),
+        ("minutes", -1, Timestamp("1969-12-31 23:59:00")),
+        ("hours", 1, Timestamp("1970-01-01 01:00:00")),
+        ("hours", -1, Timestamp("1969-12-31 23:00:00")),
+        ("days", 1, Timestamp("1970-01-02 00:00:00")),
+        ("days", -1, Timestamp("1969-12-31 00:00:00")),
+        ("weeks", 1, Timestamp("1970-01-08 00:00:00")),
+        ("weeks", -1, Timestamp("1969-12-25 00:00:00")),
+        ("months", 1, Timestamp("1970-02-01 00:00:00")),
+        ("months", -1, Timestamp("1969-12-01 00:00:00")),
+        ("years", 1, Timestamp("1971-01-01 00:00:00")),
+        ("years", -1, Timestamp("1969-01-01 00:00:00")),
+    ],
+)
+def test_dateoffset_add_sub(cases):
+    time_unit, num, expected = cases
+    offset = DateOffset(**{time_unit: num})
+    ts = Timestamp(0) + offset
+    assert ts == expected
+    ts -= offset
```

Comment:
@jbrockmendel For the question "do we have any test cases with both nano !=0 and hasattr(self, "nanoseconds")?"
When the test case is `("nanoseconds", 1, Timestamp("1970-01-01 00:00:00.000000001"))`,  running L816, thus `Timestamp("1970-01-01 00:00:00.000000001")` - `DateOffset(nanoseconds=1)` >>> nano == 1 and hasattr(self, "nanoseconds")

Reply:
Thanks.  The case I think is uncovered is where both have a nanosecond _and_ the DateOffset has a relativedelta component:

```
ts = pd.Timestamp(4)
off = pd.DateOffset(minute=2, nanoseconds=9)

>>> ts + off
Timestamp('1970-01-01 00:02:00')  # <- PR
Timestamp('1970-01-01 00:02:00.000000004')  # <- master
Timestamp('1970-01-01 00:02:00.000000013')   # <- expected
```

Reply:
Thanks and I think the PR can pass this test and let me add a new test case for this.

Reply:
The test failed, maybe I need rebase my local branch in docker from master to see why.

Reply:
> to see why

In DateOffset._apply we call _as_datetime which drops nanos, so we need to either a) not do that or b) add them back in at some point

Reply:
@jbrockmendel I read the error message carefully today, the CI failed not because of the tests. The PR could handle the nano correctly. 
The building doc process failed because of a warning below:
`warning in /home/runner/work/pandas/pandas/doc/source/whatsnew/v0.10.0.rst at block ending on line 209`

Reply:
Is there any suggestion for debugging this? I haven't figured out anything wrong in this doc.

Reply:
Rebased from master branch, 
1. The CI still failed due to `Extension Array` does not have `_data` attribute.
2. `FAILED pandas/tests/io/formats/test_to_excel.py::test_css_to_excel_good_colors[#708090-7080901].`
3.`E           FutureWarning: The pandas.SparseArray class is deprecated and will be removed from pandas in a future version. Use pandas.arrays.SparseArray instead.`

@jbrockmendel It seems that all the errors above are not related with my PR, could you take a look?